pkg/cmd/experimental/registry/registry.go

```diff
@@ -0,0 +1,267 @@
+
+If a registry service does not exist, the --create flag can be passed to
+create a deployment configuration and service that will run the registry. If you are
+running your registry in production, you should pass --replicas=2 or higher to ensure
+you have failover protection.
+
+Examples:
+  Check the default Docker registry ("docker-registry"):
+
+  $ %[1]s %[2]s
+
+  See what the registry would look like if created:
+
+  $ %[1]s %[2]s -o json
+
+  Create a registry if it does not exist with two replicas:
+
+  $ %[1]s %[2]s --create --replicas=2
+
+  Use a different registry image and see the registry configuration:
+
+  $ %[1]s %[2]s -o yaml --images=myrepo/docker-registry:mytag
+
+ALPHA: This command is currently being actively developed. It is intended to simplify
+  the tasks of setting up a Docker registry in a new installation.
+`
+
+type config struct {
+	Type          string
+	ImageTemplate variable.ImageTemplate
+	Ports         string
+	Replicas      int
+	Labels        string
+	Volume        string
+	HostMount     string
+	Create        bool
+	Credentials   string
+}
+
+const defaultLabel = "docker-registry=default"
+
+func NewCmdRegistry(f *clientcmd.Factory, parentName, name string, out io.Writer) *cobra.Command {
+	cfg := &config{
+		ImageTemplate: variable.NewDefaultImageTemplate(),
+
+		Labels:   defaultLabel,
+		Ports:    "5000",
+		Volume:   "/registry",
+		Replicas: 1,
+	}
+
+	cmd := &cobra.Command{
+		Use:   fmt.Sprintf("%s", name),
+		Short: "Install and check OpenShift Docker registry",
+		Long:  fmt.Sprintf(longDesc, parentName, name),
+
+		Run: func(cmd *cobra.Command, args []string) {
+			var name string
+			switch len(args) {
+			case 0:
+				name = "docker-registry"
+			default:
+				glog.Fatalf("No arguments are allowed to this command")
+			}
+
+			ports, err := app.ContainerPortsFromString(cfg.Ports)
+			if err != nil {
+				glog.Fatal(err)
+			}
+
+			label := map[string]string{"docker-registry": "default"}
+			if cfg.Labels != defaultLabel {
+				valid, remove, err := app.LabelsFromSpec(strings.Split(cfg.Labels, ","))
+				if err != nil {
+					glog.Fatal(err)
+				}
+				if len(remove) > 0 {
+					glog.Fatalf("You may not pass negative labels in %q", cfg.Labels)
+				}
+				label = valid
+			}
+
+			image := cfg.ImageTemplate.ExpandOrDie(cfg.Type)
+
+			namespace, err := f.OpenShiftClientConfig.Namespace()
+			if err != nil {
+				glog.Fatalf("Error getting client: %v", err)
+			}
+			_, kClient, err := f.Clients(cmd)
+			if err != nil {
+				glog.Fatalf("Error getting client: %v", err)
+			}
+
+			p, output, err := cmdutil.PrinterForCommand(cmd)
+			if err != nil {
+				glog.Fatalf("Unable to configure printer: %v", err)
+			}
+
+			generate := output
+			if !generate {
+				_, err = kClient.Services(namespace).Get(name)
+				if err != nil {
+					if !errors.IsNotFound(err) {
+						glog.Fatalf("Can't check for existing router %q: %v", name, err)
+					}
+					generate = true
+				}
+			}
+
+			if generate {
```

Comment:
Eventually we'll have deeper checks (and there is a check below).

----- Original Message -----
> > +			if err != nil {
> > +				glog.Fatalf("Unable to configure printer: %v", err)
> > +			}
> > +
> > +			generate := output
> > +			if !generate {
> > +				_, err = kClient.Services(namespace).Get(name)
> > +				if err != nil {
> > +					if !errors.IsNotFound(err) {
> > +						glog.Fatalf("Can't check for existing router %q: %v", name, err)
> > +					}
> > +					generate = true
> > +				}
> > +			}
> > +
> > +			if generate {
> 
> if !generate and exit? that will save you one indentation level
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1100/files#r25151165